### PR TITLE
Jenkins build

### PR DESCRIPTION
**Pull request by crxpandion**
depends on kafka_produce

### CODE REVIEW

File: build/packer.json [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      "type": "file",
+      "source": "build/config/",
+      "destination": "/opt/science/{{user `project`}}/config"
+    },
+    {
+      "type": "shell",
+      "inline":
+      [
+        "sudo chmod +x /opt/science/{{user `project`}}/bin/*",
+        "curl -O http://science.twitch.tv/debs/seeker_1.0.3_amd64.deb",
```
Comment (GarethLewin) [inline build/packer.json]: Better to make this a variable, then it's in one place and easier to find when you want to change it.

File: k_writer/producer.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	sp := &Producer{
+		producer:        prod,
+		topic:           topic,
+		newExpectations: make(chan *producerExpect), // this must be unbuffered
+	}
+
+	return sp, nil
+}
+
+func (p *Producer) SendMessage(key, value sarama.Encoder) error {
```
Comment (GarethLewin) [inline k_writer/producer.go]: I haven't run golint, but if I had it would complain about a func starting with Caps and no comment.

**Comment by GarethLewin** [comment]
Beyond inline comments and a general dislike of k_writer as a name, LGTM

File: k_writer/producer.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
+package k_writer
+
+import (
+	"github.com/shopify/sarama"
+)
+
+// SimpleProducer publishes Kafka messages. It routes messages to the correct broker, refreshing metadata as appropriate,
+// and parses responses for errors. You must call Close() on a producer to avoid leaks, it may not be garbage-collected automatically when
```
Comment (aasted) [inline k_writer/producer.go]: Perhaps an example of deferring would be the right thing?

Why doesn't it close and own its own client?

Comment (aasted) [inline k_writer/producer.go]: not in that order :)

Comment (aasted) [inline k_writer/producer.go]: It looks like it *does* own and close its client. Is this comment incorrect?

File: k_writer/producer.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+		topic:           topic,
+		newExpectations: make(chan *producerExpect), // this must be unbuffered
+	}
+
+	return sp, nil
+}
+
+func (p *Producer) SendMessage(key, value sarama.Encoder) error {
+	msg := &sarama.MessageToSend{Topic: p.topic, Key: key, Value: value}
+	expectation := &producerExpect{msg: msg, result: make(chan error)}
```
Comment (aasted) [inline k_writer/producer.go]: Should we set up a free list here instead of creating and deleting these all the time?

Comment (crxpandion) [inline k_writer/producer.go]: sure - thats a good optimization. Were not having memory issues right now so im not too worried about it.

File: k_writer/producer.go (hunk 3 above)
Comment (crxpandion) [inline k_writer/producer.go]: yes -opps

File: k_writer/producer.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+		select {
+		case expectation, ok := <-newExpectations:
+			if !ok {
+				delete(unmatched, nil) // let us exit when we've processed the last message
+				newExpectations = nil  // prevent spinning on a closed channel until that happens
+				continue
+			}
+			unmatched[expectation.msg] = expectation.result
+		case err := <-p.producer.Errors():
+			unmatched[err.Msg] <- err.Err
```
Comment (aasted) [inline k_writer/producer.go]: The channel is left open in this case. I'd prefer we handle nominal and error in the same way --  maybe the receiver should close the channel? Is there a preferred pattern for this sort of future?

**Comment by crxpandion** [comment]
merged with f92ff99

**Pull request closed without merging**